### ISSUE TITLE
Recommend against pattern matching with null for Unity Objects

### ISSUE THREAD

**Issue by Runaway01** [microsoft/Microsoft.Unity.Analyzers#228]
#### Problem statement
Pattern matching Unity Objects with `null` results in a false positives if the object is being destroyed. This is due to [Unity's custom equality/inequality operators for Unity Objects](https://blog.unity.com/technology/custom-operator-should-we-keep-it).
```cs
if (gameObject is null)
```
Similar false positives caused by avoiding equality/inequality operators are already covered by analyzers UNT0007, UNT0008, and UNT0023.

Using such pattern matching is even more appealing now with C#9's pattern matching improvements, specifically the `not` negation pattern.
```cs
if (gameObject is not null)
```

I haven't seen any issues from this case yet since we have only recently started using `is` for null checks, but can foresee one of these slipping through code review and causing an issue down the line.

#### Proposed solution
Any pattern matching on a Unity Object which includes `null` should be recognised by this analyzer.

The most common patterns I'd expect to see:
```cs
if (gameObject is null)
```
and
```cs
if (gameObject is not null)
```

This pattern may also appear as a 'not null' check though I suspect it's uncommon:
```cs
if (gameObject is object)
``` 
##### Suggested fix

The suggested fix would be to instead use equality/inequality operators:
```cs
if (gameObject == null)
```
and
```cs
if (gameObject != null)
```

At this time I'm not familiar with creating analysers, but hope to learn this in the future.
Thanks for taking the time to consider my suggestion 🙂

**Comment by sailro** [comment]
Yes I want to add this as well. The only technical subtility is that we are supporting VS2019 and VS2022. So currently the analyzer code is targeting our compiler infrastructure (Roslyn) v3.4.0 and I'm not sure we have everything to handle all cases.

**Comment by sailro** [comment]
Confirmed, we need Roslyn v3.7.0 for the `UnaryPatternSyntax` type and the `SyntaxKind.NotPattern` enum value.

Given the following servicing [document](https://docs.microsoft.com/fr-fr/visualstudio/releases/2019/servicing-vs2019), I think it is safe to bump the Roslyn version.

cc @jbevain what do you think?

**Comment by sailro** [comment]
All done. Thank you @Runaway01 for the raising this case.
